### DISCUSSION TITLE
Always running Global Dependencies, even when the route isn't matched - 404

### DISCUSSION THREAD

**Discussion by jonfinerty** [fastapi/fastapi#11221] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def add_request_logging_middleware_dependencies_to_request_state(
    request: Request,
    logging_dependency: Annotated[LoggingDependency, Depends(get_logging_dependency)],
) -> None:
    request.state._state[LOGGING_DEPENDENCY] = logging_dependency


class RequestLoggingMiddleware(BaseHTTPMiddleware):
    @override
    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
        request_start_time = time.time()
        response = await call_next(request)

        logging_dependency: LoggingDependency | None = request.state._state.get(LOGGING_DEPENDENCY,None)
        if logging_dependency is None:
            raise ConfigError("Uh oh, it's not here because we're in a 404")

        elapsed_time_seconds = time.time() - request_start_time

        logging_dependency.log(elapsed_time_seconds, response.status_code)

        return response
```

### Description

So, I'm trying to add some generic logging to my FastAPI application. 

The approach we have is pretty fragile:
1. A global depends on the application that constructs some dependencies we need to log (we log to a third party via a library), and puts them into the request state for access later: `request.state._state[SOME_DEP_KEY] = some_dep`
2. A middleware that after the response has been returned, logs the status code, timing etc to this third party by using the dependencies stashed in the request.

It ended up with this combination because we could use just a `Depends` for logging as we needed the response status code of the request which you don't get access to, even with a yielding depends. And we couldn't use just a middleware as we wanted to re-use some of the dependencies we use in the rest of the app.

Now this approach works decently, except in one case. If we have a request which doesn't match a route, then the global depends isn't triggered, so the dependencies aren't put in the request state and our middleware can't really work.

In short my question is, is there a way to run a Global Depends (or something similar) for requests which don't match a route/end up a 404?

Any ideas? Thanks!

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

3.12.2
